extended ssm_eltwise_mul

kpaigwar opened the pull request:
Extend ssm_eltwise_mul kernel to handle a case when the input a is already of shape [1, 1, B, 2E*N]

input a shape : [1, 1, 32, 32*5120]
input b shape : [1, 1, 32, 5120]

Here we just need to repeat_interleave input b and do tile by tile eltwise mul with input a.

Check the green portion in the diagram which we have considered
![image](https://github.com/tenstorrent/tt-metal/assets/132708568/fc591a3a-989b-48ed-9a12-7ffa83b8d70f)

TT-BrianLiu reviewed (changes requested):
If B is just repeated W times, why does our kernel code even need to change? Just read in one tile and reuse that for all of compute just like how it was before. 

Also in general, code like this is bad for couple of reasons:
```
 noc_async_read_tile((i*32 + in0_block_id_per_in1_block)*(in0_num_blocks>1), s0, l1_write_addr_in0);
 ```
 - You should `ifdef` (or `constexpr`) the old/new paths so we don't regress the other use case (this applies to compute as well)
 - You shouldn't have multiplies like that. You can just increment the counter.
 - You should use `constexpr` variables instead of hard-coded values.

kpaigwar commented:
> * You should `ifdef` (or `constexpr`) the old/new paths so we don't regress the other use case (this applies to compute as well)
* In order to generalize the ssm_eltwise_mul kernel for other input shapes, we thought to modify the same reader and writer kernels.
* To avoid if else conditions for two cases I used arithmetic conditions like this `(in0_num_blocks>1)` that covers both the case 
a. when same tile needs to be read and multiplied
b. when input is already repeated, read tile by tile

TT-BrianLiu commented:
> > * You should `ifdef` (or `constexpr`) the old/new paths so we don't regress the other use case (this applies to compute as well)
> 
> My bad, input a can also be a result of previous `ssm_eltwise_mul` making it of shape [1, 1, B, 2E*N] which means each tile can different.
> 
> * In order to generalize the ssm_eltwise_mul kernel for other input shapes, we thought to modify the same reader and compute kernels.
> * To avoid if else conditions for two cases I used arithmetic conditions like this `(in0_num_blocks>1)` that covers both the case
>   a. when input has only one tile, read the same tile again and again. [This makes CBs a bit bigger but design remains the same]
>   b. when input is already repeated or shape of 2E*N, read tile by tile

For optimized ops like these, I don't think you want to regress perf for other use case. You need to make `in0_num_blocks>1` is compile time anyways, so you should `ifdef` the code paths. It's still general but more optimal for both cases.

kpaigwar commented:
> > > * You should `ifdef` (or `constexpr`) the old/new paths so we don't regress the other use case (this applies to compute as well)
> > 
> > 
> > My bad, input a can also be a result of previous `ssm_eltwise_mul` making it of shape [1, 1, B, 2E*N] which means each tile can different.
> > 
> > * In order to generalize the ssm_eltwise_mul kernel for other input shapes, we thought to modify the same reader and compute kernels.
> > * To avoid if else conditions for two cases I used arithmetic conditions like this `(in0_num_blocks>1)` that covers both the case
> >   a. when input has only one tile, read the same tile again and again. [This makes CBs a bit bigger but design remains the same]
> >   b. when input is already repeated or shape of 2E*N, read tile by tile
> 
> For optimized ops like these, I don't think you want to regress perf for other use case. `in0_num_blocks>1` is compile time anyways, so you should `ifdef` the code paths. It's still general but more optimal for both cases.

Got it, I am OOO today, will make the changes when back

kpaigwar committed:
#8106: use #ifdef to handle ssm_eltwise_mul cases

kpaigwar commented:
@TT-BrianLiu  I have addressed your comments. lmk if this what you were expecting.
All the tests are passing with the expected device perf of 0.07 ms in both the cases.

kpaigwar commented:
```
============================================================================================ short test summary info =============================================================================================
PASSED tests/tt_eager/python_api_testing/unit_testing/misc/test_ssm_eltwise_mul.py::test_ssm_eltwise_mul[32-5120-dtype0-in0_mem_config0-in1_mem_config0-out_mem_config0]
PASSED tests/tt_eager/python_api_testing/unit_testing/misc/test_ssm_eltwise_mul.py::test_ssm_eltwise_mul[32-5120-dtype0-in0_mem_config0-in1_mem_config0-out_mem_config1]
PASSED tests/tt_eager/python_api_testing/unit_testing/misc/test_ssm_eltwise_mul.py::test_ssm_eltwise_mul[32-5120-dtype0-in0_mem_config0-in1_mem_config1-out_mem_config0]
PASSED tests/tt_eager/python_api_testing/unit_testing/misc/test_ssm_eltwise_mul.py::test_ssm_eltwise_mul[32-5120-dtype0-in0_mem_config0-in1_mem_config1-out_mem_config1]
PASSED tests/tt_eager/python_api_testing/unit_testing/misc/test_ssm_eltwise_mul.py::test_ssm_eltwise_mul[32-5120-dtype0-in0_mem_config1-in1_mem_config0-out_mem_config0]
PASSED tests/tt_eager/python_api_testing/unit_testing/misc/test_ssm_eltwise_mul.py::test_ssm_eltwise_mul[32-5120-dtype0-in0_mem_config1-in1_mem_config0-out_mem_config1]
PASSED tests/tt_eager/python_api_testing/unit_testing/misc/test_ssm_eltwise_mul.py::test_ssm_eltwise_mul[32-5120-dtype0-in0_mem_config1-in1_mem_config1-out_mem_config0]
PASSED tests/tt_eager/python_api_testing/unit_testing/misc/test_ssm_eltwise_mul.py::test_ssm_eltwise_mul[32-5120-dtype0-in0_mem_config1-in1_mem_config1-out_mem_config1]
PASSED tests/tt_eager/python_api_testing/unit_testing/misc/test_ssm_eltwise_mul.py::test_ssm_eltwise_mul[163840-5120-dtype0-in0_mem_config0-in1_mem_config0-out_mem_config0]
PASSED tests/tt_eager/python_api_testing/unit_testing/misc/test_ssm_eltwise_mul.py::test_ssm_eltwise_mul[163840-5120-dtype0-in0_mem_config0-in1_mem_config0-out_mem_config1]
PASSED tests/tt_eager/python_api_testing/unit_testing/misc/test_ssm_eltwise_mul.py::test_ssm_eltwise_mul[163840-5120-dtype0-in0_mem_config0-in1_mem_config1-out_mem_config0]
PASSED tests/tt_eager/python_api_testing/unit_testing/misc/test_ssm_eltwise_mul.py::test_ssm_eltwise_mul[163840-5120-dtype0-in0_mem_config0-in1_mem_config1-out_mem_config1]
PASSED tests/tt_eager/python_api_testing/unit_testing/misc/test_ssm_eltwise_mul.py::test_ssm_eltwise_mul[163840-5120-dtype0-in0_mem_config1-in1_mem_config0-out_mem_config0]
PASSED tests/tt_eager/python_api_testing/unit_testing/misc/test_ssm_eltwise_mul.py::test_ssm_eltwise_mul[163840-5120-dtype0-in0_mem_config1-in1_mem_config0-out_mem_config1]
PASSED tests/tt_eager/python_api_testing/unit_testing/misc/test_ssm_eltwise_mul.py::test_ssm_eltwise_mul[163840-5120-dtype0-in0_mem_config1-in1_mem_config1-out_mem_config0]
PASSED tests/tt_eager/python_api_testing/unit_testing/misc/test_ssm_eltwise_mul.py::test_ssm_eltwise_mul[163840-5120-dtype0-in0_mem_config1-in1_mem_config1-out_mem_config1]
PASSED tests/tt_eager/python_api_testing/unit_testing/misc/test_ssm_eltwise_mul.py::test_ssm_eltwise_mul_with_program_cache
============================================================================================== 17 passed in 13.86s ===============================================================================================
                 Device | INFO     | Closing user mode device drivers
```

If this looks good, I will also add support for the third case where inputs are of shape `(1, 1, 32, 32)` and `(1, 1, 32, 32*5120)` respectively

TT-BrianLiu reviewed (approved):
Looks good!

kpaigwar commented:
> Looks good!

I have also implemented the third case where where input a and input b are of shape [1, 1, 32, 32] and [1, 32, 32*5120] respectively. Should I push in the same PR?
